alf/agents/a2c/a2c_agent.py

```diff
@@ -0,0 +1,194 @@
+            time_step_spec,
+            action_spec,
+            policy=policy,
+            collect_policy=policy,
+            train_sequence_length=None,
+            debug_summaries=debug_summaries,
+            summarize_grads_and_vars=summarize_grads_and_vars,
+            train_step_counter=train_step_counter)
+
+    def _initialize(self):
+        pass
+
+    def _train(self, experience, weights=None):
+        time_steps = ts.TimeStep(experience.step_type,
+                                 experience.reward,
+                                 experience.discount,
+                                 experience.observation)
+
+        with tf.GradientTape() as tape:
+            loss_info = self._loss(time_steps, experience.action, weights)
+        grad_variables = self._value_net.variables + self._actor_net.variables
+        grads = tape.gradient(loss_info.loss, grad_variables)
+        self._apply_gradients(grads, grad_variables, self._optimizer)
+        self.train_step_counter.assign_add(1)
+        return tf.nest.map_structure(tf.identity, loss_info)
+
+    def _calc_values_and_returns(self, time_steps):
+        values = self._value_net(time_steps.observation)[0]
+        final_value = values[:, -1]
+        final_is_last = time_steps.is_last()[:, -1]
+        final_value = final_value * tf.cast(~final_is_last, tf.float32)
+        discounts = time_steps.discount[:, :-1]
+        discounts *= tf.cast(~time_steps.is_last(), tf.float32)[:, 1:]
+        discounts *= self._gamma
+        rewards = time_steps.reward[:, :-1]
+        returns = value_ops.discounted_return(
+            rewards,
+            discounts=discounts,
+            final_value=final_value,
+            time_major=False)
+        if len(returns.shape) != len(final_value.shape):
+            final_value = tf.reshape(final_value, (-1, 1))
+        returns = tf.concat((returns, final_value), axis=1)
+        if self._debug_summaries:
+            self._summary('Infos', [
+                ('values', values),
+                ('returns', returns)])
+        return values, returns
+
+    def _loss(self, time_steps, actions, weights):
+        values, returns = self._calc_values_and_returns(time_steps)
+        batch_size = time_steps.discount.shape.dims[0]
+        policy_state = self._collect_policy.get_initial_state(batch_size=batch_size)
+        actions_distribution = self.collect_policy.distribution(
+            time_steps, policy_state).action
+        action_log_prob = common.log_probability(
+            actions_distribution, actions,
+            self.action_spec)
+        valid_mask = tf.cast(~time_steps.is_last(), tf.float32)
+        action_log_prob *= valid_mask
+        advantage = tf.stop_gradient(returns - values)
+        advantage *= valid_mask
+        policy_loss = -tf.reduce_sum(action_log_prob * advantage, axis=-1)
+        policy_loss = tf.reduce_mean(policy_loss)
+
+        td_error = self._td_errors_loss_fn(values, returns)
```

Comment:
Since returns[-1] - values[-1] == 0 and advantage[-1]==0(because line 126), td_error[-1] , the last time step is not used for calculating loss, which is correct. But I think it's better to explicit them from the loss calculation (by removing actions[-1] from actions) so that it's easier to understand.